native/protocol/src/x3dh.rs

```diff
@@ -0,0 +1,364 @@
+        recipient_ik.to_montgomery().to_bytes(),
+    ));
+    let dh3 = ek.diffie_hellman(&spk.pre_key);
+
+    let sk = match opk {
+        Some(one_time_prekey) => {
+            let dh4 = ek.diffie_hellman(&one_time_prekey);
+            kdf(&[
+                dh1.to_bytes(),
+                dh2.to_bytes(),
+                dh3.to_bytes(),
+                dh4.to_bytes(),
+            ]
+            .concat())
+        }
+        None => kdf(&[dh1.to_bytes(), dh2.to_bytes(), dh3.to_bytes()].concat()),
+    };
+
+    Ok(X3DHSendKeyAgreement {
+        ek: X25519PublicKey::from(&ek),
+        sk,
+    })
+}
+
+// Alice then sends Bob an initial message containing:
+//    Alice's identity key IKA
+//    Alice's ephemeral key EKA
+//    Identifiers stating which of Bob's prekeys Alice used
+//    An initial ciphertext encrypted with some AEAD encryption scheme [4] using AD as associated data and using an encryption key which is either SK or the output from some cryptographic PRF keyed by SK.
+pub fn initiate_send(
+    prekey_bundle: PreKeyBundle,
+    sender_identity: String,
+    sender_ik: &SigningKey,
+    message: &[u8],
+) -> Result<([u8; 32], Message), X3DHError> {
+    let X3DHSendKeyAgreement { ek, sk } = initiate_send_get_sk(
+        prekey_bundle.ik,
+        prekey_bundle.spk,
+        prekey_bundle.opk,
+        sender_ik,
+    )?;
+    // Alice then calculates an "associated data" byte sequence AD that contains identity information for both parties:
+    //   AD = Encode(IKA) || Encode(IKB)
+    // Alice may optionally append additional information to AD, such as Alice and Bob's usernames, certificates, or other identifying information.
+    let associated_data = [
+        sender_ik.verifying_key().to_bytes(),
+        prekey_bundle.ik.to_bytes(),
+    ]
+    .concat();
+
+    // The initial ciphertext is typically the first message in some post-X3DH communication protocol. In other words, this ciphertext typically has two roles, serving as the first message within some post-X3DH protocol, and as part of Alice's X3DH initial message.
+    // After sending this, Alice may continue using SK or keys derived from SK within the post-X3DH protocol for communication with Bob
+    let ciphertext = encrypt_data(
+        Payload {
+            msg: message,
+            aad: &associated_data,
+        },
+        &ChaCha20Poly1305::new_from_slice(&sk).unwrap(),
+    )?;
+
+    Ok((
+        sk,
+        Message {
+            sender_identity,
+            sender_ik: sender_ik.verifying_key(),
+            ek,
+            opk: prekey_bundle.opk,
+            ciphertext,
+        },
+    ))
+}
+
+fn initiate_recv_get_sk(
+    sender_ik: &VerifyingKey,
+    ek: X25519PublicKey,
+    opk: Option<X25519StaticSecret>,
+    receiver_ik: &SigningKey,
+    spk: &X25519StaticSecret,
+) -> [u8; 32] {
+    let dh1 = spk.diffie_hellman(&X25519PublicKey::from(sender_ik.to_montgomery().to_bytes()));
```

Comment:
I get that single lines are nice,  but JFC. You're making 3 function calls and making a macro refernece

Reply:
Montgomery is [this badass dude](https://en.wikipedia.org/wiki/Peter_Montgomery_(mathematician)). Do you want fast math or slow math? Montgomery multiplication is fast and you need to convert your key into a form that works with it.